File: lib/domain/cultivation_plan/mappers/plan_save_crop_task_template_attributes_mapper.rb
# frozen_string_literal: true

module Domain
  module CultivationPlan
    module Mappers
      class PlanSaveCropTaskTemplateAttributesMapper
        # @param link_row [Dtos::PublicPlanSaveCropTaskTemplateLinkRow]
        # @param task_row [Dtos::PublicPlanSaveAgriculturalTaskReferenceRow]
        # @param user_task_name [String, nil] persisted user AgriculturalTask name (wins over reference template)
        # @return [Hash]
        def self.attributes_for_create(link_row:, task_row:, user_task_name: nil)
          {
            name: user_task_name.presence || link_row.name.presence || task_row.name,
            description: link_row.description.nil? ? task_row.description : link_row.description,
            time_per_sqm: link_row.time_per_sqm.nil? ? task_row.time_per_sqm : link_row.time_per_sqm,
            weather_dependency: link_row.weather_dependency.nil? ? task_row.weather_dependency : link_row.weather_dependency,
            required_tools: tools_for_create(link_row, task_row),
            skill_level: link_row.skill_level.nil? ? task_row.skill_level : link_row.skill_level,
            task_type: link_row.task_type.nil? ? task_row.task_type : link_row.task_type,
            task_type_id: link_row.task_type_id.nil? ? task_row.task_type_id : link_row.task_type_id,
            is_reference: link_row.is_reference
          }
        end

        def self.tools_for_create(link_row, task_row)
          tools = link_row.required_tools.nil? ? task_row.required_tools : link_row.required_tools
          return [] if tools.nil?

          Array(tools).dup
        end
        private_class_method :tools_for_create
      end
    end
  end
end

File: crates/agrr-domain/src/cultivation_plan/mappers/plan_save_crop_task_template_attributes_mapper.rs
//! Ruby: `Domain::CultivationPlan::Mappers::PlanSaveCropTaskTemplateAttributesMapper`

use std::collections::BTreeMap;

use serde_json::{json, Value};

use crate::cultivation_plan::dtos::{
    PublicPlanSaveAgriculturalTaskReferenceRow, PublicPlanSaveCropTaskTemplateLinkRow,
};

pub fn attributes_for_create(
    link_row: &PublicPlanSaveCropTaskTemplateLinkRow,
    task_row: &PublicPlanSaveAgriculturalTaskReferenceRow,
    user_task_name: Option<&str>,
) -> BTreeMap<String, Value> {
    let name = user_task_name
        .filter(|s| !s.is_empty())
        .or(link_row.name.as_deref().filter(|s| !s.is_empty()))
        .or(task_row.name.as_deref())
        .map(str::to_string);

    BTreeMap::from([
        ("name".into(), json!(name)),
        (
            "description".into(),
            json!(link_row
                .description
                .clone()
                .or_else(|| task_row.description.clone())),
        ),
        (
            "time_per_sqm".into(),
            json!(link_row.time_per_sqm.or(task_row.time_per_sqm)),
        ),
        (
            "weather_dependency".into(),
            json!(link_row
                .weather_dependency
                .clone()
                .or_else(|| task_row.weather_dependency.clone())),
        ),
        (
            "required_tools".into(),
            json!(tools_for_create(link_row, task_row)),
        ),
        (
            "skill_level".into(),
            json!(link_row.skill_level.clone().or_else(|| task_row.skill_level.clone())),
        ),
        (
            "task_type".into(),
            json!(link_row.task_type.clone().or_else(|| task_row.task_type.clone())),
        ),
        (
            "task_type_id".into(),
            json!(link_row.task_type_id.or(task_row.task_type_id)),
        ),
        ("is_reference".into(), json!(link_row.is_reference)),
    ])
}

fn tools_for_create(
    link_row: &PublicPlanSaveCropTaskTemplateLinkRow,
    task_row: &PublicPlanSaveAgriculturalTaskReferenceRow,
) -> Vec<String> {
    link_row
        .required_tools
        .clone()
        .or_else(|| task_row.required_tools.clone())
        .unwrap_or_default()
}
